### PR TITLE
Fix handling of expired cookies so they are not stored in CookieJar

### PR DESCRIPTION
**Pull request by alandtse**
## What do these changes do?

Properly schedule expiration date for cookies that are provided to the CookieJar with a date in the past which will result in such cookies being immediately expired by `_do_expiration()`.  Currently, the expiration date is set to the `timestamp()` (i.e., unix timestamp) which will make almost all values far in the future since it's compared to the `event_loop.time()` which increments from the start of the loop.

Marked WIP because this fix breaks the tests in `tests/test_cookiejar.py` for `test_constructor` and `TestCookieJarSafe.test_expires`.  I believe there are incorrect assumptions about behavior but I wanted to verify the test assumptions before making changes to the test suite. Mainly, should expired cookies be stored in the CookieJar and just never sent in requests? Or should they automatically be deleted?
The current test suite assumes expired cookies will not be deleted automatically.

I can complete the PR and the remaining steps once I understand the architecture.

## Are there changes in behavior for the user?

CookieJar will automatically reject expired cookies and not store in the CookieJar
## Related issue number

#4063 
## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [x] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names. 
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

### CODE REVIEW

File: aiohttp/cookiejar.py [hunk 1]
```diff
@@ -175,7 +175,9 @@ def update_cookies(self,
                 if expires:
                     expire_time = self._parse_date(expires)
                     if expire_time:
-                        self._expire_cookie(expire_time.timestamp(),
+                        diff = (expire_time.timestamp() -
```
Comment (asvetlov) [inline aiohttp/cookiejar.py]: From my understanding the first parameter of `_expire_cookie` is called `when` and it is an absolute time (seconds since UXIN epoch), not a delta with the current time.
I like to keep the parameter meaning.

Comment (alandtse) [reply aiohttp/cookiejar.py]: I don't believe that's a true assumption for `_expire_cookie`. Look at line 168 where it's used for max-age expiration. It's a time relative to the loop_time().  

The first parameter is the sum of the `self._loop.time()` + seconds left from max-age. It's a relative time and not tied to UNIX epoch because `loop.time()` is [isn't tied to the UNIX epoch](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.time).  Because `loop.time()` isn't tied to the UNIX epoch, the correct answer is to pass some value <= 0 which will cause an immediate expiration.  You could include some additional math functions to force it to be 0 if you don't want it to be negative, but that isn't necessary.

Comment (alandtse) [reply aiohttp/cookiejar.py]: If it's supposed to be `time.time()`, then I can easily change my edit to `time.time()+diff` for consistency.  Will that resolve your concern?  However, line 86 in `do_expiration()` will break because you're comparing `loop.time` which will always be much smaller than `time.time()`.  To resolve that I could just do a `loop.time() + diff`, which will also typically be a very negative number (again since unix epochs are so much larger).

For the test suite, I still need an answer on whether CookieJar should allow storage of expired cookies as that is a core assumption in the test suite.

Comment (asvetlov) [reply aiohttp/cookiejar.py]: I mean using time.time() consistently in `CookieJar` for representing absolute time looks like a proper fix. It doesn't need `loop.time()` at all, at least seems so.

Comment (alandtse) [reply aiohttp/cookiejar.py]: Yes, I think that looks right. Do you want me to make that switch throughout?

Comment (alandtse) [reply aiohttp/cookiejar.py]: It can be done with datetime unless you want to import time too.

Comment (asvetlov) [reply aiohttp/cookiejar.py]: Please do. `datetime` is even better, we never mix absolute and relative time again.

Comment (alandtse) [reply aiohttp/cookiejar.py]: Ok, what about the test suite assumption?

Comment (asvetlov) [reply aiohttp/cookiejar.py]: Do you mean tests that become fail after the code change?
Tests should be updated.

**Comment by mjpieters** [comment]
There are client functional tests in [`test_client_functional.py`](https://github.com/aio-libs/aiohttp/blob/master/tests/test_client_functional.py), including

https://github.com/aio-libs/aiohttp/blob/dc39183d4351c2fc87dbfa9bcf938567d754571b/tests/test_client_functional.py#L1959-L1983

I'd copy that test as `test_set_cookies_expired` and verify that expired cookies are handled correctly.

File: aiohttp/cookiejar.py (hunk 1 above)
Comment (mjpieters) [reply aiohttp/cookiejar.py]: There are indeed more issues here. If `datetime` is desired then I'd rework the time components in the whole class and use `datetime.datetime` instances with UTC time throughout.

* `MAX_TIME` should be `datetime.datetime.max.replace(tzinfo=datetime.timezone.utc)`
* `self._next_expiration = ceil(self._loop.time())` in the `CookieJar` `__init__` `_clear()`  and  should be set to `datetime.datetime.now(datetime.timezone.utc).replace(microsecond=0) + datetime.timedelta(seconds=0)` (so microseconds cleared and a second added to align with the next whole second).
* The `_do_expirations()` method should switch to use `now = datetime.datetime.now(datetime.timezone.utc)` and `self._next_expiration = next_expiration.replace(microsecond=0) + datetime.timedelta(seconds=1)`
* alter `_expire_cookie` to make `when` a `datetime` instance:

    ```python
    def _expire_cookie(self, when: datetime.datetime, domain: str, name: str) -> None:
        self._next_expiration = min(self._next_expiration, when)
        self._expirations[(domain, name)] = when
    ```
* Change the `max-age` handling to pass in `datetime.datetime.now(timezone.utc) + delta_seconds`.

Technically speaking `datetime.datetime.now(datetime.timezone.utc).replace(microsecond=0) + datetime.timedelta(seconds=0)` is not quite the same thing as `ceil(self._loop.time())` if you *happen* to get a `datetime.datetime()` instance with `microsecond` already set to 0, but I don't think it is worth explicitly testing for that here.

It is probably worth creating a helper function for it:

```python
def _next_whole_second() -> datetime.datetime:
    """The current time rounded up to the next whole second"""
    return (
        datetime.datetime.now(datetime.timezone.utc).replace(microsecond=0) +
        datetime.timedelta(seconds=0)
    )
```

Comment (alandtse) [reply aiohttp/cookiejar.py]: Ok I've implemented this; time to work on the tests and make sure I haven't broken anything.

**Comment by alandtse** [comment]
Can I get this reviewed please? Looking at the failed tests it's either related to 3.5 syntax which I believe you are deprecating or components I didn't touch. `aiohttp\tests\test_worker.py`

Please note I added one dev dependency for freezegun because mocking datetime is a pain because it's a builtin.

File: aiohttp/cookiejar.py [hunk 2]
```diff
@@ -48,15 +47,16 @@ class CookieJar(AbstractCookieJar):
...
 
     def __init__(self, *, unsafe: bool=False) -> None:
         self._loop = get_running_loop()
         self._cookies = defaultdict(SimpleCookie)  #type: DefaultDict[str, SimpleCookie]  # noqa
         self._host_only_cookies = set()  # type: Set[Tuple[str, str]]
         self._unsafe = unsafe
-        self._next_expiration = ceil(self._loop.time())
-        self._expirations = {}  # type: Dict[Tuple[str, str], int]
+        self._next_expiration: datetime.datetime = self._next_whole_second()
+        self._expirations: Dict[Tuple[str, str], datetime.datetime] = {}
```
Comment (mjpieters) [inline aiohttp/cookiejar.py]: Again, keep the hints in comments. 

```suggestion
        self._expirations = {}  # type: Dict[Tuple[str, str], datetime.datetime]
```

File: aiohttp/cookiejar.py [hunk 3]
```diff
@@ -48,15 +47,16 @@ class CookieJar(AbstractCookieJar):
 
     DATE_YEAR_RE = re.compile(r"(\d{2,4})")
 
-    MAX_TIME = 2051215261.0  # so far in future (2035-01-01)
+    MAX_TIME: datetime.datetime = datetime.datetime.max.replace(
```
Comment (mjpieters) [inline aiohttp/cookiejar.py]: Don’t use variable annotations, this is the 3.5 syntax error you see. Put type hints in a comment:

```python 
MAX_TIME = datetime.datetime.max.replace(
    tzinfo=datetime.timezone.utc)  # type: datetime.datetime
```

I don’t think a type hint is needed here however; dropping it is fine:

```suggestion
    MAX_TIME = datetime.datetime.max.replace(
```

File: aiohttp/cookiejar.py [hunk 4]
```diff
@@ -82,8 +82,16 @@ def __iter__(self) -> 'Iterator[Morsel[str]]':
     def __len__(self) -> int:
         return sum(1 for i in self)
 
+    def _next_whole_second(self) -> datetime.datetime:
```
Comment (mjpieters) [inline aiohttp/cookiejar.py]: This helper function should just be that: a function at module level.

File: aiohttp/cookiejar.py [hunk 5]
```diff
@@ -82,8 +82,16 @@ def __iter__(self) -> 'Iterator[Morsel[str]]':
...
+        """Return current time rounded up to the next whole second."""
+        return (
+            datetime.datetime.now(
+                datetime.timezone.utc).replace(microsecond=0) +
+            datetime.timedelta(seconds=0)
+        )
+
     def _do_expiration(self) -> None:
-        now = self._loop.time()
+        now: datetime.datetime = datetime.datetime.now(datetime.timezone.utc)
```
Comment (mjpieters) [inline aiohttp/cookiejar.py]: No hint needed. Mypy can see what the function returns. 
```suggestion
        now = datetime.datetime.now(datetime.timezone.utc)
```

File: aiohttp/cookiejar.py [hunk 6]
```diff
@@ -102,12 +110,16 @@ def _do_expiration(self) -> None:
         for key in to_del:
             del expirations[key]
 
-        self._next_expiration = ceil(next_expiration)
+        try:
+            self._next_expiration = (next_expiration.replace(microsecond=0) +
+                                     datetime.timedelta(seconds=1))
+        except OverflowError:
+            self._next_expiration = self.MAX_TIME
```
Comment (mjpieters) [inline aiohttp/cookiejar.py:109]: I like how you handled the possible overflow here :-)

File: aiohttp/cookiejar.py [hunk 7]
```diff
@@ -165,8 +177,10 @@ def update_cookies(self,
             if max_age:
                 try:
                     delta_seconds = int(max_age)
-                    self._expire_cookie(self._loop.time() + delta_seconds,
-                                        domain, name)
+                    self._expire_cookie(datetime.datetime.now(
+                        datetime.timezone.utc) +
+                        datetime.timedelta(seconds=delta_seconds),
+                        domain, name)
```
Comment (mjpieters) [inline aiohttp/cookiejar.py]: You need to handle OverflowError here. A malicious or broken server could send max_age=_more seconds than time available in the universe_.

File: aiohttp/cookiejar.py [hunk 8]
```diff
@@ -175,8 +189,9 @@ def update_cookies(self,
                 if expires:
                     expire_time = self._parse_date(expires)
                     if expire_time:
-                        self._expire_cookie(expire_time.timestamp(),
-                                            domain, name)
+                        self._expire_cookie(datetime.datetime.fromtimestamp(
+                            expire_time.timestamp(), tz=datetime.timezone.utc),
+                            domain, name)
```
Comment (mjpieters) [inline aiohttp/cookiejar.py]: The `expire_time` value here is a perfectly serviceable UTC value already, why convert again?

File: aiohttp/cookiejar.py [hunk 9]
```diff
@@ -48,15 +47,16 @@ class CookieJar(AbstractCookieJar):
...
+        tzinfo=datetime.timezone.utc)
 
     def __init__(self, *, unsafe: bool=False) -> None:
         self._loop = get_running_loop()
         self._cookies = defaultdict(SimpleCookie)  #type: DefaultDict[str, SimpleCookie]  # noqa
         self._host_only_cookies = set()  # type: Set[Tuple[str, str]]
         self._unsafe = unsafe
-        self._next_expiration = ceil(self._loop.time())
-        self._expirations = {}  # type: Dict[Tuple[str, str], int]
+        self._next_expiration: datetime.datetime = self._next_whole_second()
```
Comment (mjpieters) [inline aiohttp/cookiejar.py]: Drop the hint here

```suggestion
        self._next_expiration = self._next_whole_second()
```

File: requirements/dev.txt [hunk 10]
```diff
@@ -1,3 +1,4 @@
 -r ci.txt
 -r towncrier.txt
 cherry_picker==1.3.2; python_version>="3.6"
+freezegun==0.3.12
```
Comment (mjpieters) [inline requirements/dev.txt]: This should be added to ci.txt I think, not here. Appveyor breaks because this is the wrong location for the dependency.

File: aiohttp/cookiejar.py (hunk 2 above)
Comment (alandtse) [reply aiohttp/cookiejar.py]: Ok, I thought the plan was to deprecate 3.5, but will revert.

File: requirements/dev.txt (hunk 10 above)
Comment (alandtse) [reply requirements/dev.txt]: Moved.

File: aiohttp/cookiejar.py (hunk 4 above)
Comment (alandtse) [reply aiohttp/cookiejar.py]: Moved.

File: aiohttp/cookiejar.py (hunk 6 above)
Comment (alandtse) [reply aiohttp/cookiejar.py:109]: Thanks.

File: aiohttp/cookiejar.py [hunk 11]
```diff
@@ -48,15 +47,16 @@ class CookieJar(AbstractCookieJar):
...
 
     def __init__(self, *, unsafe: bool=False) -> None:
         self._loop = get_running_loop()
         self._cookies = defaultdict(SimpleCookie)  #type: DefaultDict[str, SimpleCookie]  # noqa
         self._host_only_cookies = set()  # type: Set[Tuple[str, str]]
         self._unsafe = unsafe
-        self._next_expiration = ceil(self._loop.time())
-        self._expirations = {}  # type: Dict[Tuple[str, str], int]
+        self._next_expiration = self._next_whole_second()
+        self._expirations = {}  # type: Dict[Tuple[str, str], datetime.datetime]
```
Comment (mjpieters) [inline aiohttp/cookiejar.py]: This line is now too long for flake8. Line 56 had the same issue and used a noqa flag to tell flake8 to ignore it. We could do the same here but I’d prefer it to be specific:

```suggestion
        self._expirations = {}  # type: Dict[Tuple[str, str], datetime.datetime] # noqa: E501
```

@asvetlov remind me to talk to you about flake8-bugbear and the flexible line length overflow setting. :-) if this project is to use black in future that’s a good addition anyway.

Comment (alandtse) [reply aiohttp/cookiejar.py]: Change made before I saw the suggestion. It's also the reason I went to inline type hints. ;)

**Comment by alandtse** [comment]
Decided to just add the test for max-age overflow since it wasn't hard.

File: tests/test_client_functional.py [hunk 12]
```diff
@@ -2032,6 +2033,30 @@ def dumps(obj):
...
 
+async def test_set_cookies_max_age_overflow(aiohttp_client) -> None:
+
+    async def handler(request):
+        ret = web.Response()
+        ret.headers.add('Set-Cookie',
+                        'overflow=overflow; '
+                        'HttpOnly; Path=/'
+                        " Max-Age=" +
+                        str(int(datetime.datetime.max.timestamp())) + "; ")
```
Comment (mjpieters) [inline tests/test_client_functional.py]: 👍

Comment (mjpieters) [reply tests/test_client_functional.py]: Unfortunately, the very act of asking for the time stamp can trigger an overflow exception, based on the current timezone! See the [appveyor test failures](https://ci.appveyor.com/project/aio-libs/aiohttp/builds/27616968/job/4i6eq555evqijl56#L3419):

```python
ERROR    aiohttp.server:web_protocol.py:380 Error handling request
Traceback (most recent call last):
  File "C:\projects\aiohttp\aiohttp\web_protocol.py", line 405, in _handle_request
    resp = await self._request_handler(request)
  File "C:\projects\aiohttp\aiohttp\web_app.py", line 330, in _handle
    resp = await handler(request)
  File "C:\projects\aiohttp\tests\test_client_functional.py", line 2044, in handler
    str(int(datetime.datetime.max.timestamp())) + "; ")
OverflowError: timestamp out of range
```

`datetime.max` Is a naive datetime value and has no timezone but a timestamp is always a value relative to 1970-01-01 midnight UTC, so Python uses the local timezone to interpret naive values when converting. In timezones west from UTC that requires adding hours to the datetime value, and so we get an overflow here.

Comment (alandtse) [reply tests/test_client_functional.py]: Oops. I guess I just need to choose a static large int value to avoid that issue.

**Comment by alandtse** [comment]
I honestly don't know why the Travis build failed.

Edit: Nevermind, figured it out.

**Comment by mjpieters** [comment]
Looks great to me; @asvetlov, @webknjaz, any remarks?
